File: app/src/main/java/com/stedi/randomimagegenerator/app/model/data/generatorparams/base/SimpleIntegerParams.java
package com.stedi.randomimagegenerator.app.model.data.generatorparams.base;

import android.os.Parcel;

public abstract class SimpleIntegerParams extends GeneratorParams {
    private Integer value;

    public SimpleIntegerParams() {
        if (!canBeRandom())
            value = 1;
    }

    public Integer getValue() {
        return value;
    }

    public void setRandomValue() {
        if (canBeRandom())
            this.value = null;
    }

    public void setValue(int value) {
        if (value < 1)
            throw new IllegalArgumentException("value must be > 0");
        this.value = value;
    }

    public abstract boolean canBeRandom();

    @Override
    public boolean isEditable() {
        return true;
    }

    @Override
    public boolean equals(Object o) {
        if (!super.equals(o)) return false;

        SimpleIntegerParams that = (SimpleIntegerParams) o;

        return value != null ? value.equals(that.value) : that.value == null;
    }

    @Override
    public int hashCode() {
        return value != null ? value.hashCode() : 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(this.value);
    }

    protected SimpleIntegerParams(Parcel in) {
        this.value = (Integer) in.readValue(Integer.class.getClassLoader());
    }
}

File: app/src/main/java/com/stedi/randomimagegenerator/app/model/data/generatorparams/base/SimpleIntegerParams.kt
package com.stedi.randomimagegenerator.app.model.data.generatorparams.base

import android.os.Parcel
import android.support.annotation.CallSuper
import com.j256.ormlite.field.DatabaseField

abstract class SimpleIntegerParams : GeneratorParams {

    @Suppress("LeakingThis")
    @DatabaseField(columnName = "integer_value")
    private var value: Int? = if (canBeRandom()) null else 1

    // OrmLite required
    constructor()

    fun setRandomValue() {
        if (canBeRandom()) {
            this.value = null
        }
    }

    fun setValue(value: Int) {
        if (value < 1) {
            throw IllegalArgumentException("value must be > 0")
        }
        this.value = value
    }

    fun getValue(): Int? = value

    abstract fun canBeRandom(): Boolean

    @CallSuper
    override fun equals(other: Any?): Boolean {
        if (!super.equals(other)) return false
        val that = other as SimpleIntegerParams
        return if (value != null) value == that.value else that.value == null
    }

    @CallSuper
    override fun hashCode() = value?.hashCode() ?: 0

    @CallSuper
    override fun writeToParcel(dest: Parcel, flags: Int) {
        super.writeToParcel(dest, flags)
        dest.writeValue(this.value)
    }

    protected constructor(parcel: Parcel) : super(parcel) {
        this.value = parcel.readValue(Int::class.java.classLoader) as Int?
    }
}
